src/item_location.cpp

```diff
@@ -0,0 +1,182 @@
+#include "item_location.h"
+
+#include "enums.h"
+#include "game.h"
+#include "map.h"
+#include "Character.h"
+#include "player.h"
+#include "vehicle.h"
+#include <map>
+
+int item_location::get_inventory_position()
+{
+    return INT_MIN;
+}
+
+item_is_null::item_is_null()
+{
+    what = nullptr;
+}
+
+void item_is_null::remove_item()
+{ }
+
+item *item_is_null::get_item()
+{
+    return nullptr;
+}
+
+
+item_on_map::item_on_map( const tripoint &p, const item *which )
+{
+    location = p;
+    what = nullptr;
+    // Ensure the item exists where we want it
+    const auto items = g->m.i_at( p );
+    for( auto &i : items ) {
+        if( &i == which ) {
+            what = &i;
+        }
+    }
+}
+
+void item_on_map::remove_item()
+{
+    if( what == nullptr ) {
+        return;
+    }
+
+    g->m.i_rem( location, what );
+    what = nullptr;
+}
+
+item *item_on_map::get_item()
+{
+    if( what == nullptr ) {
+        return nullptr;
+    }
+
+    auto items = g->m.i_at( location );
+    for( auto &i : items ) {
+        if( &i == what ) {
+            return &i;
+        }
+    }
+
+    return nullptr;
+}
+
+
+item_on_person::item_on_person( Character &ch, const item *which )
+{
+    who = &ch;
+    auto the_item = ch.items_with( [which]( const item &it ) {
+        return &it == which;
+    } );
+
+    if( !the_item.empty() ) {
+        what = the_item[0];
+    } else {
+        what = nullptr;
+    }
+}
+
+void item_on_person::remove_item()
+{
+    if( what == nullptr ) {
+        return;
+    }
+
+    who->remove_items_with( [this]( const item &it ) {
+        return &it == what;
+    } );
+
+    what = nullptr;
+}
+
+item *item_on_person::get_item()
+{
+    if( what == nullptr ) {
+        return nullptr;
+    }
+
+    auto items = who->items_with( [this]( const item &it ) {
+        return &it == what;
+    } );
+
+    if( !items.empty() ) {
+        return items[0];
+    } else {
+        what = nullptr;
```

Comment:
This should also show a debug message as it indicates a very serious problem: the item somebody tried to use has vanished unexpectedly.